src/control/lib/hardware/hwloc/provider.go

```diff
@@ -15,6 +15,43 @@ import (
 	"github.com/daos-stack/daos/src/control/logging"
 )
 
+type ctxKey string
+
+const (
+	topoKey    ctxKey = "hwlocTopology"
+	cleanupKey ctxKey = "hwlocFreeTopology"
+)
+
+// CacheContext adds a cache of the hwloc topology to the provided context.
+func CacheContext(parent context.Context, log logging.Logger) (context.Context, error) {
+	topo, cleanup, err := NewProvider(log).initTopology()
+	if err != nil {
+		return parent, err
+	}
+
+	topoCtx := context.WithValue(parent, topoKey, topo)
+	return context.WithValue(topoCtx, cleanupKey, cleanup), nil
+}
+
+// Cleanup frees the hwloc cache in the context, if any.
+func Cleanup(ctx context.Context) error {
+	cleanup, ok := ctx.Value(cleanupKey).(func())
```

Comment:
I think this function should probably be idempotent -- i.e. not an error to call it with a context that doesn't have the cleanup key. Just becomes a no-op. Especially given that the common pattern will be for this to be called in a defer where the error isn't checked anyhow... Unless you can see a situation where that would cause a problem.

Reply:
That makes sense to me.